llvm/lib/Target/DirectX/DXILOpBuilder.cpp

```diff
@@ -249,17 +309,95 @@ static FunctionType *getDXILOpFunctionType(const OpCodeProperty *Prop,
+
+  return -1;
+}
+
 namespace llvm {
 namespace dxil {
 
+// No extra checks on TargetTripleStr need be performed to verify that the
+// Triple is well-formed or that the target is supported since these checks
+// would have been done at the time the module M is constructed in the earlier
+// stages of compilation.
+DXILOpBuilder::DXILOpBuilder(Module &M, IRBuilderBase &B)
+    : M(M), B(B), TargetTripleStr(M.getTargetTriple()) {}
+
 CallInst *DXILOpBuilder::createDXILOpCall(dxil::OpCode OpCode, Type *ReturnTy,
                                           Type *OverloadTy,
                                           SmallVector<Value *> Args) {
+
+  auto Major = Triple(TargetTripleStr).getDXILVersion().getMajor();
+  auto MinorOrErr = Triple(TargetTripleStr).getDXILVersion().getMinor();
+  uint32_t Minor = MinorOrErr.has_value() ? *MinorOrErr : 0;
+  VersionTuple DXILVer(Major, Minor);
+  // Get Shader Stage Kind
+  Triple::EnvironmentType ShaderEnv = Triple(TargetTripleStr).getEnvironment();
+
   const OpCodeProperty *Prop = getOpCodeProperty(OpCode);
+  int OlIndex = getPropIndex(Prop->Overloads, DXILVer);
+  uint16_t ValidTyMask = Prop->Overloads[OlIndex].ValidTys;
 
   OverloadKind Kind = getOverloadKind(OverloadTy);
-  if ((Prop->OverloadTys & (uint16_t)Kind) == 0) {
-    report_fatal_error("Invalid Overload Type", /* gen_crash_diag=*/false);
+
+  // Check if the operation supports overload types and OverloadTy is valid
+  // per the specified types for the operation
+  if ((ValidTyMask != OverloadKind::UNDEFINED) &&
+      (ValidTyMask & (uint16_t)Kind) == 0) {
+    report_fatal_error(Twine("Invalid Overload Type for DXIL operation - ") +
+                           getOpCodeName(OpCode),
+                       /* gen_crash_diag=*/false);
+  }
+
+  // Ensure Environment type is known
+  if (ShaderEnv == Triple::UnknownEnvironment) {
+    report_fatal_error(
+        Twine(DXILVer.getAsString()) +
+            ": Unknown Compilation Target Shader Stage specified ",
+        /*gen_crash_diag*/ false);
+  }
+
+  // Perform necessary checks to ensure Opcode is valid in the targeted shader
+  // kind
+  int StIndex = getPropIndex(Prop->Stages, DXILVer);
+  uint16_t ValidShaderKindMask = Prop->Stages[StIndex].ValidStages;
+  enum ShaderKind ModuleStagekind = getShaderKindEnum(ShaderEnv);
+
+  // Ensure valid shader stage properties are specified
+  if (ValidShaderKindMask == ShaderKind::removed) {
+    report_fatal_error(
+        Twine(DXILVer.getAsString()) +
+            ": Unsupported Target Shader Stage for DXIL operation - " +
+            getOpCodeName(OpCode),
+        /*gen_crash_diag*/ false);
+  }
+
+  // Shader stage need not be validated since getShaderKindEnum() fails
+  // for unknown shader stage.
+
+  // Verify the target shader stage is valid for the DXIL operation
+  if (!(ValidShaderKindMask & ModuleStagekind)) {
+    auto ShaderEnvStr = Triple(TargetTripleStr).getEnvironmentName();
```

Comment:
You have a bunch of places that do this `Triple(TargetTripleStr)` over and over again. Should you just cache the constructed triple?
```suggestion
    StringRef ShaderEnvStr = Triple(TargetTripleStr).getEnvironmentName();
```
```suggestion
    auto ShaderEnvStr = Triple(TargetTripleStr).getEnvironmentName();
```

Reply:
> You have a bunch of places that do this `Triple(TargetTripleStr)` over and over again. Should you just cache the constructed triple?

Replaced `std::string DXILOpBuilder::TargetTripleStr` with `Triple::TargetTriple` to cache the Triple.